### ISSUE TITLE
`match` generates unhelpful log message about non-linear patterns

### ISSUE THREAD

**Issue by florence** [racket/racket#1386]
In one of my programs I get the log message "non-linear pattern used in `match` with ... at" (which seems to be logged with `log-error`).

The docs don't mention anything about non-linear patterns, so I'm not sure what I should do about it (or if I should care about it at all.)

**Comment by rfindler** [comment]
bump

**Comment by rfindler** [comment]
@samth ?

**Comment by samth** [comment]
Oh, sorry. A non-linear pattern is one where the same variable is used multiple times, as in:

```
(match e
  [(list x x) "xs are equal"])
```

Those do not work correctly in general with `...` but can't be removed entirely for backwards compatibility reasons. I'll add documentation explaining what they are.

**Comment by rfindler** [comment]
Thanks, @samth 

Perhaps also add something to the log message so people who see that first know what to do?

**Comment by lexi-lambda** [comment]
This came up again just now in [this mailing list post](https://groups.google.com/d/msg/racket-users/fkOL2Rp9Hnw/-rBxsrbyCwAJ).

**Comment by alexgian** [comment]
> This came up again just now in this mailing list post.

Yes, indeed it did.  My two bits worth:
You can't **not** have this feature or "leave it out".  IMHO it's a pretty key pattern-matching capability, and would make the difference between me using the library or not.

I haven't looked at the source, nor do I have enough mastery of syntax to help out, but is the problem in the original "Andrew Wright" code, which I understand is historically the precursor of this library, or was it introduced with the port to Racket?

**Comment by samth** [comment]
#2231 is a PR that will make things better here. A few things here.

1. It's obviously not ideal to warn about things instead of producing an error. The last time I tested, making this an error would not be backwards compatible, but I'm testing again.
2. This has never worked -- when non-linear patterns were added I didn't know that this didn't work, leading the current problem. This also doesn't work (because it's inexpressible) in the pattern matching systems of other programming languages I know of.
3. Making this work would require significant changes to the generalized-sequences implementation. I'm not sure that's a good idea, in particular because I'm not sure how to do it while keeping the current structure of the implementation. The pattern matcher in `redex` supports this but works very differently. Of course, if someone has an implementation idea that addresses this I'd be very interested.

**Comment by samth** [comment]
Fixed in https://github.com/racket/racket/pull/5467.

**Comment by mfelleisen** [comment]
@samth -- It turns out that this change (disallowing non-linear patterns) affected 7 or 8 large oprivate code bases of mine. Yes, I know they are private, but it makes no sense making the public as packages (2 books).  

Can I complain now or is it too late? 

To make these code bases operational again is going to be major work

**Comment by samth** [comment]
Yes, definitely we can reconsider things here. But I will want to see what the problematic patterns are, because the current state is very broken and likely they aren't doing anything sensible right now.

**Comment by mfelleisen** [comment]
So I have scripts that (1) run the code in book chapters and (2) that translate this code into type-set rules. One of the patterns works best for running the code, the other one is for the type-setting script because it makes more sense to type-set it this way.  

Here is an example: 

```
      [(state 'S e s (prog* '() '() r))
       (match-define (and z* (list z* ...)) (all-variables-of-r r))
       (match-define (and v* (list v* ...)) (value-of* z* e s))
       (define r^ (rename-all-variables-of-r-to-unique-names r))
       (match-define (and x* (list x* ...)) (all-variables-of-r^ r^))
       (match-define (and xl* (list xl* ...)) (allocate-new-location* s (length x*)))
       (define e0 ρ)
       (define e1 (extend* e0 x* xl*))
       (define s1 (extend* s xl* v*))
       (define k (prog* '() '() r))
       (state r^ e1 s1 (pop k))]
```

I have dozens if not a couple of hundred such patterns in my code.  

(FWIW, I understand the problems with non-linear patterns.  I ruled them out too when I added `match` to Scheme 84.)

**Comment by mfelleisen** [comment]
(P.S. Right now I am working around this problem with your amazing rackup (though using it wasn't completely pain free). Thanks for that work!!! )

**Comment by samth** [comment]
Are they all of the form `(and x (list x ...))`? What do you think the semantics of that should be?

**Comment by mfelleisen** [comment]
Mostly.  (I recall trying another non-linear pattern in some places but I wrote these things last summer and I am only resuming work now. Perhaps I can figure out a `grep` command to find things.) 

`(and x (list x ...))` should match a list and bind `x` to the list.

**Comment by samth** [comment]
So you were looking for the same behavior as `(and x (list _ ...))` or just `(list x ...)` both of which do what you describe.

Unfortunately that interpretation is not really compatible with saying that `(and 3 (list 3 ...))` should match nothing at all. I guess `(shared ([x (list x x x x x)]) x)` would make sense for your original pattern. 

I could come up with a semantics like: "if a variable appears non-linearly under ..., then each individual pattern has an equality constraint, but if it also appears outside the ..., then the equality constraint is with the whole list", but that feels rather ugly.

**Comment by rfindler** [comment]
Another option might be to allow some specific patterns just for backwards compatibility reasons? Something more specific than what you write above, say requiring the same variable in multiple places?

**Comment by samth** [comment]
@rfindler well the whole thing is about the same variable in multiple places :) 

I guess we could allow that specific pattern as a special case and document it as a weird exception.

**Comment by mfelleisen** [comment]
The `(and x (list _ ...))` gave me an idea.  It is possible that can salvage the situation by fixing just one place in the type-setting script.  I'll try this tomorrow. (Fixing dozens if not a couple hundred places would bother me.) 

(Once again, I am glad I run the bleeding edge release.)

**Comment by mfelleisen** [comment]
I spent an hour fixing the code bases that used `match-define` with variable-repeating patterns.  Both of them run on 'bleeding edge' racket now, so I am fine. I am closing the issue again. Sorry for the delay.

**Comment by samth** [comment]
Were they all of the form `(and x (list x ...))`?

**Comment by mfelleisen** [comment]
Almost all. (I had created a function to type-set other shapes in a slightly different way, and that helped.)

**Comment by samth** [comment]
What did those patterns look like?

**Comment by mfelleisen** [comment]
They ‘buried’ variables in calls to other constructors or predicates. Say 

```
(match '(1 2 3)
    [(and (list a b c) (cons a rst)) b])
```

**Comment by samth** [comment]
That example doesn't involve `...`, though. I would really appreciate seeing the actual code that you had that no longer worked.

**Comment by mfelleisen** [comment]
Ah, I had no other … pattern.  That was the point of having a case split in the type setter.
